server/proof_fetcher.go

```diff
@@ -1,53 +1,155 @@
+	r.rehashNode = storage.Node{Hash: node.Hash}
+	r.rehashing = true
+}
+
+func (r *rehasher) endRehashing() {
+	if r.rehashing {
+		r.proof = append(r.proof, &trillian.Node{NodeHash: r.rehashNode.Hash})
+		r.rehashing = false
+	}
+}
+
+func (r *rehasher) rehashedProof(leafIndex int64) (trillian.Proof, error) {
+	r.endRehashing()
+	return trillian.Proof{LeafIndex: leafIndex, ProofNode: r.proof}, r.proofError
+}
+
+// dedupAndFetchNodes() removes duplicates from the set of fetches and then passes the result to
+// storage. After writng this code I realised I don't have a solid proof that fetches for Merkle
+// paths could involve duplicate nodes so it could be that this isn't ever useful. Further
+// thought is required.
+func dedupAndFetchNodes(tx storage.NodeReader, treeRevision int64, fetches []merkle.NodeFetch) ([]storage.Node, error) {
+	// To start with we remove any duplicate fetches
+	m := make(map[string]storage.NodeID)
+	proofNodeIDs := make([]storage.NodeID, 0, len(fetches))
+
+	// Remove duplicates, preserving the order of the input otherwise
+	for _, fetch := range fetches {
+		if _, ok := m[fetch.NodeID.String()]; !ok {
+			// First time we've seen the ID
+			m[fetch.NodeID.String()] = fetch.NodeID
+			proofNodeIDs = append(proofNodeIDs, fetch.NodeID)
 		}
 	}
 
+	if len(proofNodeIDs) < len(fetches) {
+		glog.V(2).Infof("deduplication saved %d node fetch(es)", len(fetches)-len(proofNodeIDs))
+	}
+
+	// Use the deduplicated list of nodeIDs in the storage fetch
 	proofNodes, err := tx.GetMerkleNodes(treeRevision, proofNodeIDs)
 	if err != nil {
-		return trillian.Proof{}, err
+		return []storage.Node{}, err
 	}
 
 	if len(proofNodes) != len(proofNodeIDs) {
-		return trillian.Proof{}, fmt.Errorf("expected %d nodes in proof but got %d", len(proofNodeIDs), len(proofNodes))
+		return []storage.Node{}, fmt.Errorf("expected %d nodes from storage but got %d", len(proofNodeIDs), len(proofNodes))
 	}
 
-	proof := make([]*trillian.Node, 0, len(proofNodeIDs))
-	for i, node := range proofNodes {
-		// additional check that the correct node was returned
-		if !node.NodeID.Equivalent(proofNodeIDs[i]) {
-			return trillian.Proof{}, fmt.Errorf("expected node %v at proof pos %d but got %v", proofNodeIDs[i], i, node.NodeID)
-		}
+	// Then rebuild what we would have got if we'd fetched the duplicates
+	nodes := make([]storage.Node, 0, len(fetches))
+	nm := make(map[string]storage.Node)
+
+	for _, node := range proofNodes {
+		nm[node.NodeID.String()] = node
+	}
 
-		idBytes, err := proto.Marshal(node.NodeID.AsProto())
-		if err != nil {
-			return trillian.Proof{}, err
+	for _, fetch := range fetches {
+		node, ok := nm[fetch.NodeID.String()]
+
+		if !ok {
+			return []storage.Node{}, fmt.Errorf("wanted node ID: %s but storage didn't return it", fetch.NodeID.String())
 		}
 
-		proof = append(proof, &trillian.Node{NodeId: idBytes, NodeHash: node.Hash, NodeRevision: node.NodeRevision})
+		nodes = append(nodes, node)
+	}
+
+	for i, node := range nodes {
```

Comment:
As I mentioned we have got a lot of checks that things have behaved correctly. They've saved us from blowing up logs a couple of times.